### PR TITLE
Verify crates are publish-able on CI

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit updates our CI to verify that all crates are publish-able at
all times on every commit. During the 0.19.0 release we found another
case where the crates as they live in this repository weren't
publish-able, so the hope is that this no longer comes up again!

The script added in this commit also takes the time/liberty to remove
the existing bump/publish scripts and instead replace them with one Rust
script originally sourced from wasm-bindgen. The intention of this
script is that it has three modes:

* `./publish bump` - bumps version numbers which are sent as a PR to get
  reviewed (probably with a changelog as well)

* `./publish verify` - run on CI on every commit, builds every crate we
  publish as if it's being published to crates.io, notably without raw
  access to other crates in the repository.

* `./publish publish` - publishes all crates to crates.io, passing the
  `--no-verify` flag to make this a much speedier process than it is
  today.

### CODE REVIEW

**Comment by alexcrichton** [comment]
The intention of this commit is to make the release process much more seamless, and also require less waiting on whomever is doing the release. Since all validation happens on CI now publication is simply sending a bunch of `*.crate` files to crates.io to get published now.

File: scripts/publish.rs [hunk 1]
```diff
@@ -0,0 +1,318 @@
+//! Helper script to publish the wasmtime and cranelift suites of crates
+//!
+//! Usage:
+//!
+//! * First, compile this script
+//! * Next, set cwd to the root of the wasmtime repository
+//! * Execute `./publish bump` to bump versions
+//! * Send a PR
+//! * Merge when green
+//! * Execute `./publish publish` to publish crates
```
Comment (sunfishcode) [inline scripts/publish.rs:10]: This usage comment is kind of a less-complete version of the docs in contributing-release-process.md. Would it make sense to just link to that document, and then make the comment here just briefly describe the CLI options?

File: scripts/publish.rs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    let major = iter.next().expect("major version");
+    let minor = iter.next().expect("minor version");
+    let patch = iter.next().expect("patch version");
+    if major != 0 {
+        format!("{}.0.0", major + 1)
+    } else if minor != 0 {
+        format!("0.{}.0", minor + 1)
+    } else {
+        format!("0.0.{}", patch + 1)
+    }
```
Comment (sunfishcode) [inline scripts/publish.rs:240]: I may be missing something here, but I don't see what this function is supposed to be doing.

Comment (alexcrichton) [reply scripts/publish.rs:245]: I added some more documentation to this, but the intention is to basically figure out the next version number we're going to write down everywhere for each crate individually.

Comment (sunfishcode) [reply scripts/publish.rs:245]: Ah, makes sense now. Thanks!

**Pull request merged**